connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala

```diff
@@ -460,27 +460,120 @@ private class KeyValueGroupedDatasetImpl[K, V, IK, IV](
   }
 
   override protected def aggUntyped(columns: TypedColumn[_, _]*): Dataset[_] = {
-    // TODO(SPARK-43415): For each column, apply the valueMap func first...
+    val originalDs = sparkSession.newDataset(ivEncoder, plan)
+
+    // Apply the value transformation, get a DS of two columns "iv" and "v".
+    // If any of "iv" or "v" consists of a single primitive field, wrap it with a struct so it
+    // would not be flattened.
+    val valueTransformedDf = applyValueMapFunc(originalDs)
+
+    // Restore column/field names, if this KVGDS is from a DF.
+    val (valueTransformedDfWithSchema, ivFields, vFields) =
```

Comment:
I don't understand why we need to retain the names for multiple column values? For a single column I understand that it is needed.